### ISSUE TITLE
This plugin's transforms run before config file transforms

### ISSUE THREAD

**Issue by pauleveritt** [11ty/eleventy-img#228]
My config file extends JavaScript Templates to allow JSX/TSX. It uses `config.addTransform` to render JSX objects into a string.

Everything works fine until I add the `eleventy-img` plugin:

```
  eleventyConfig.addExtension(["11ty.jsx", "11ty.ts", "11ty.tsx"], {
    key: "11ty.js",
  });

  eleventyConfig.addTransform("tsx", async (content: any) => {
    const result = await renderToString(content);
    return `<!doctype html>\n${result}`;
  });

  eleventyConfig.addPlugin(eleventyImageTransformPlugin, {
    formats: ["avif", "png"],
  });
```

Its HTML transform executes before mine. Since it is given an object instead of a string, it returns the string `[object Object]` and my transform no longer has JSX to work on.

The [docs say](https://www.11ty.dev/docs/transforms/): "Transforms added via plugins are inserted via the [second configuration stage for plugins](https://www.11ty.dev/docs/plugins/#creating-a-plugin)." I believe `eleventy-img` is doing its transform before other transforms can step in.

**Comment by zachleat** [comment]
You’re right that the HTML transformer used by the Image plugin (and the input path to url plugin and the HTML `<base>` plugin) will run before application-level transforms. Filed https://github.com/11ty/eleventy/issues/3297 for this.

But—would this would work better for `tsx` files?

![image](https://github.com/11ty/eleventy-img/assets/39355/9d642597-27e8-4ab6-8d35-9115135d9971)

via slide 60 on https://www.zachleat.com/web/11tyconf-keynote-2024/

**Comment by pauleveritt** [comment]
@zachleat When I tried that, I got an error saying I couldn't add a `compile` when extending a builtin.

**Comment by zachleat** [comment]
A ha — I see what you mean. Investigating further!

**Comment by zachleat** [comment]
Woops, incorrect commit reference, added support for more complex addExtension definitions (e.g. `compile`) when extending builtins here: https://github.com/11ty/eleventy/commit/ff78badba3dae78696dd4ce9b032f4dd122b5b68

This means you’ll be able to do things like this:

```js
eleventyConfig.addExtension(["11ty.jsx", "11ty.ts", "11ty.tsx"], {
	key: "11ty.js",
	read: false,
	compile: function() {
		return async function(data) {
			let content = await this.defaultRenderer(data);
			return renderToString(content);
		};
	}
});
```

Shipping with 3.0.0-alpha.11.

**Comment by zachleat** [comment]
Moving this to core to correctly milestone the solution here.
